static/js/tracking.ts

```diff
@@ -0,0 +1,231 @@
+
+    // Explore page buttons
+    "explore_page_adventure",
+    "explore_page_level"
+];
+const CLICK_COUNTS = "clickCounts";
+const LAST_ACTIVE = "lastActiveTime";
+const INTERVAL_KEY = "interval";
+
+
+let clickCounts: any = [];
+let lastActiveTime = Date.now();
+let changesSent = false;
+
+
+export function initializeTracking() {
+    document.addEventListener("DOMContentLoaded", documentLoaded);
+}
+
+
+function documentLoaded() {
+    // attach events
+    document.addEventListener('click', trackEvent);
+    document.addEventListener('change', trackEvent);
+    
+    
+    // initialize variables
+    clickCounts = handleLocalStorage(CLICK_COUNTS);
+    lastActiveTime = handleLocalStorage(LAST_ACTIVE, Date.now());
+    
+    if (isLoggedIn()) {
+        removeTrackingInterval(); // Possibly removing lingering ones.
+        setTrackingInterval(); // Resume with fresh timer
+    }
+}
+
+function removeTrackingInterval() {
+    const storedData = localStorage.getItem(INTERVAL_KEY);
+    if (storedData) {
+        try {
+            const parsedData = JSON.parse(storedData);
+            clearInterval(parsedData.id); // Clear any potentially lingering timer
+            // console.log(parsedData.id, " interval was removed")
+        } catch (error) {
+            console.error("Error parsing tracking interval data:", error);
+        }
+    }
+
+}
+
+function setTrackingInterval() {
+  const timerId = setInterval(checkUserActivity, WAITING_TIME);
+  localStorage.setItem(INTERVAL_KEY, JSON.stringify({ id: timerId, timestamp: Date.now() }));
+}
+
+
+async function trackEvent(event: Event) {
+    // the following check is necessary since some elements issue click and change events.
+    const currentTime = Date.now();
+    const inactiveDuration = currentTime - lastActiveTime;
+    if (inactiveDuration <= 200) {
+        return;
+    }
+    const target = event.target as HTMLElement;
+    // console.log(target, event.type)
+    if (target.matches('button') || target.matches('a') || target.matches('input') || target.matches('select') || target.matches("div")) {
+        let elementIdOrName = target.id;
+
+        if (!elementIdOrName && target.hasAttribute("name")) {
+            elementIdOrName = target.getAttribute("name") || "";
+        }
+
+        if (ELEMENT_TO_TRACK.includes(elementIdOrName)) {
+            clickCounts = handleLocalStorage(CLICK_COUNTS);
+            
+            const page = window.location.pathname;
+
+            const value = (target as HTMLInputElement).value || "";
+
+            clickCounts.push({time: currentTime, id: elementIdOrName, page, extra: value});
+            
+            // console.log(target, clickCounts)
+            // console.log(`Event: ${event.type}, Element ID or Name: ${elementIdOrName}, Click Count: ${clickCounts}`);
+            handleUserActivity(clickCounts);
+        } 
+    }
+}
+
+// Function to handle user activity
+function handleUserActivity(clickCounts: any) {
```

Comment:
Please take a couple of moments to define a type for the data you're storing, so you don't have to use `any`. The compiler will be able to help you better, and it will be more clear to other people later what information you're storing and why:

```
interface ClickEvent {
  time: number,
  id: string,
  page: string,
  extra: any,
}
```